cgosqlite/cgosqlite.go

```diff
@@ -272,8 +272,12 @@ func (stmt *Stmt) ColumnTableName(col int) string {
 	return C.GoString((*C.char)(unsafe.Pointer(C.sqlite3_column_table_name(stmt.stmt.ptr(), C.int(col)))))
 }
 
-func (stmt *Stmt) Step() (row bool, err error) {
-	res := C.ts_sqlite3_step(stmt.stmt.int())
+func (stmt *Stmt) Step(colType []sqliteh.ColumnType) (row bool, err error) {
+	var ptr *C.char
+	if len(colType) > 0 {
```

Comment:
Could be `!= 0` (saves an instruction on x86!)

Reply:
That's not true after the Go compiler's SSA phase and all its rando rules.

Like: IIRC it already has a rule that `len` always returns something non-negative (in the portable SSA rules). And then an x86-specific SSA rule about the thing you just said.

So it applies the two and gets the same result.

You can verify at https://godbolt.org/ comparing,

```
func foo(b []byte) {
    if len(b) > 0 {
        panic("hi")
    }
}
```

vs

```
func foo(b []byte) {
    if len(b) != 0 {
        panic("hi")
    }
}
```

same output on that `len` line:

```
        TESTQ   BX, BX
        JNE     foo_pc40
```

Reply:
Hah, and if you write `len(b) > -1` that deadcodes away and doesn't generate any code at all, as it compiles to `if true {` effectively.

Reply:
Interesting. I did it using `go tool compile`, and got a compare and a branch vs. a single CBZ, it's quite possible the actual codegen would rewrite the pair. (Stipulated, that even without checking it this isn't really a concern)

Reply:
I also just did it with `go tool compile` and see the same as godbolt.

Were you using a slice? Maybe you weren't hitting the `len(S) is always non-negative` SSA rule?

Reply:
I don't have the code anymore but I think I just used an int, which would explain why it couldn't apply that rule (careless of me).

I promise my goal wasn't to nerdsnipe you 😂